### PR TITLE
Add functinal tests for bitfield

### PR DESCRIPTION
**Pull request by melonedo**
I built a simple C library to test against Clang.jl's bitfield implementation. The results seem fine.

### CODE REVIEW

File: test/test_bitfield.jl [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    build_dir = joinpath(@__DIR__, "build", "bitfield")
+
+    config_cmd = `$cmake -B $build_dir -S $src_dir`
+    if Sys.WORD_SIZE == 32
+        if Sys.iswindows()
+            config_cmd = `$config_cmd -A win32`
+        elseif Sys.islinux()
+            config_cmd = `$config_cmd -D CMAKE_C_FLAGS=-m32 -D CMAKE_CXX_FLAGS=-m32`
+            # Not actually working, so fail fast
+            @info "Can not build libbitfield on 32-bit linux"
```
Comment (Gnimuc) [inline test/test_bitfield.jl:16]: Just out of curiosity. Why does this test not love 32-bit Linux?

**Comment by melonedo** [comment]
Without these flags, CI builds x64 on x86; with these flags CI can not even compile anything. So I disable it until someone figures out how to do that.



---Original---
From: "Yupei ***@***.***&gt;
Date: Tue, Jan 4, 2022 17:49 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [JuliaInterop/Clang.jl] Add functinal tests for bitfield (PR #367)




 
@Gnimuc commented on this pull request.
 
 
In test/test_bitfield.jl:
 &gt; +using CMake +using Clang.Generators + +function build_libbitfield() +    @info "Building libbitfield. This will break if CMake fails to find a C compiler" +    src_dir = joinpath(@__DIR__, "bitfield") +    build_dir = joinpath(@__DIR__, "build", "bitfield") + +    config_cmd = `$cmake -B $build_dir -S $src_dir` +    if Sys.WORD_SIZE == 32 +        if Sys.iswindows() +            config_cmd = `$config_cmd -A win32` +        elseif Sys.islinux() +            config_cmd = `$config_cmd -D CMAKE_C_FLAGS=-m32 -D CMAKE_CXX_FLAGS=-m32` +            # Not actually working, so fail fast +            @info "Can not build libbitfield on 32-bit linux"  
Just out of curiosity. Why does this test not love 32-bit Linux?
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by Gnimuc** [comment]
An alternative way is to use BinayBuilder.jl for building that little library.

**Comment by melonedo** [comment]
BinaryBuilder does not offer a easy solution for building on current platform only, and I think downloading toolkit and building for every platform seems like too much overhead on a local test machine.

File: test/test_bitfield.jl [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        build_dir = joinpath(@__DIR__, "build", "bitfield")
+
+        config_cmd = `$cmake -B $build_dir -S $src_dir`
+        if Sys.WORD_SIZE == 32
+            if Sys.iswindows()
+                config_cmd = `$config_cmd -A win32`
+            elseif Sys.islinux()
+                config_cmd = `$config_cmd -D CMAKE_C_FLAGS=-march=native -D CMAKE_CXX_FLAGS=-march=native`
+            end
+        end
```
Comment (Gnimuc) [inline test/test_bitfield.jl:19]: I'm wondering why CMake can not handle this for us.

**Comment by melonedo** [comment]
These flags are useless in the sense that whether with these flags or not CMake will fail because "gcc can not compile a trivial program" with the combination of flags.
Maybe the linux-32bit mirror does not have a working gcc for its own architecture.



---Original---
From: "Yupei ***@***.***&gt;
Date: Mon, Jan 10, 2022 13:54 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [JuliaInterop/Clang.jl] Add functinal tests for bitfield (PR#367)




 
@Gnimuc commented on this pull request.
 
 
In test/test_bitfield.jl:
 &gt; +        if Sys.WORD_SIZE == 32 +            if Sys.iswindows() +                config_cmd = `$config_cmd -A win32` +            elseif Sys.islinux() +                config_cmd = `$config_cmd -D CMAKE_C_FLAGS=-march=native -D CMAKE_CXX_FLAGS=-march=native` +            end +        end  
I'm wondering why CMake can not handle this for us.
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by melonedo** [comment]
It turns out that BinaryBuilder is less accessible than native tool though, it only works on ubuntu-x64 platform.

File: test/bitfield/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+products = [
+    LibraryProduct(["libbitfield"], :libbitfield)
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = Dependency[
+]
+
+# Build the tarballs, and possibly a `build.jl` as well.
+build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies)
```
Comment (Gnimuc) [inline test/bitfield/build_tarballs.jl:35]: BinaryBuilder is not supposed to be used in this way. It's for building and deploying binaries in a controlled environment. After building the binary, you need to deploy it somewhere by running `julia build_tarballs.jl --deploy=.....`, and then use the deployed JLL package for testing. 

As this is a library merely for testing Clang.jl, so there is no need to use Yggdrasil for deploying.

**Comment by Gnimuc** [comment]
I'm merging this as it may not be worth the hassle. 32bit systems will quickly fade into history.

**Pull request merged**